### DISCUSSION TITLE
FMEmbedding

### DISCUSSION THREAD

**Discussion by kuzma-long** [RUCAIBox/RecBole#1257] [General] [answered]
您好，关于特征embedding，我想知道FeatureType中的'token'、'float'、'token_seq'是分别如何做embedding的，如果是评论数据是要分词吗？

**Comment by Sherry-XLL** [comment, marked as answer]
@kuzma-long 您好！

RecBole 支持四种数据格式，'token'、'token_seq'、'float'、'float_seq'。

<div class="section" id="format" style="box-sizing: border-box;"><div class="wy-table-responsive" style="box-sizing: border-box; margin-bottom: 24px; max-width: 100%; overflow: auto;">

feat_type | Explanations | Examples
-- | -- | --
token | single discrete feature | user_id, age
token_seq | discrete features sequence | review
float | single continuous feature | rating, timestamp
float_seq | continuous feature sequence | vector

</div></div><div class="section" id="examples" style="box-sizing: border-box;"></h2>


一个 `token` 可以直接作为一个 ID 用于特征 embedding，`float` 是浮点式的具体数据，一般直接用于数值的相关使用，无需 embedding；而 `float_seq` 通常用于加载预训练好的 `embedding` 向量。

由于 `token_seq` 类的数据本身就是以分隔符分割的 `token`  序列，无论是 `token` 还是 `token_seq`，RecBole 都会将字段映射为连续的 ID，可以通过 `field2id_token` 和 `field2token_id` 进行 ID 的相互转换：

```
field2id_token (dict): Dict mapping feature name (str) to a :class:`np.ndarray`, which stores the original token
    of this feature. For example, if ``test`` is token-like feature, ``token_a`` is remapped to 1, ``token_b``
    is remapped to 2. Then ``field2id_token['test'] = ['[PAD]', 'token_a', 'token_b']``. (Note that 0 is
    always PADDING for token-like features.)

field2token_id (dict): Dict mapping feature name (str) to a dict, which stores the token remap table
    of this feature. For example, if ``test`` is token-like feature, ``token_a`` is remapped to 1, ``token_b``
    is remapped to 2. Then ``field2token_id['test'] = {'[PAD]': 0, 'token_a': 1, 'token_b': 2}``.
    (Note that 0 is always PADDING for token-like features.)
```

也就是说，若为 `token_seq` 类的评论数据，RecBole 会将评论序列中的所有 `token` 都映射为统一的内部 ID，然后通过 `rnn_utils.pad_sequence` 填充到统一的长度。


以 `ml-100k.item` 为例，电影的 `item_id` 和发行年份 `release_year` 为 `token` 数据，电影名称 `movie_title` 和电影类型 `class` 为 `token_seq` 数据，原始的原子文件为： 

https://github.com/RUCAIBox/RecBole/blob/896155c56fb2f426554d1179dae85dc0a8f4a38e/dataset/ml-100k/ml-100k.item#L1-L6

经过内部映射转换为 `Interaction` 数据格式之后，每一列特征都变为了 `torch.int64` 类型的 `Tensor` 序列：

```
The batch_size of interaction: 1683
    item_id, torch.Size([1683]), cpu, torch.int64
    movie_title, torch.Size([1683, 15]), cpu, torch.int64
    release_year, torch.Size([1683]), cpu, torch.int64
    class, torch.Size([1683, 6]), cpu, torch.int64
```

此时映射后的数据变为了：

<div class="wy-table-responsive" style="box-sizing: border-box; margin-bottom: 24px; max-width: 100%; overflow: auto; caret-color: rgb(64, 64, 64); color: rgb(64, 64, 64); font-family: Lato, proxima-nova, &quot;Helvetica Neue&quot;, Arial, sans-serif; font-size: 16px; font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-size-adjust: auto; -webkit-text-stroke-width: 0px; text-decoration: none;">

item_id:token | movie_title:token_seq | release_year:token | genre:token_seq
-- | -- | -- | --
25 | [1, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] | 1 | [1, 2, 3, 0, 0, 0]
148 | [3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] | 1 | [4, 5, 6, 0, 0, 0]
234 | [4, 5, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] | 1 | [6, 0, 0, 0, 0, 0]
48 | [6, 7, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] | 1 | [4, 3, 7, 0, 0, 0]
76 | [8, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] | 1 | [8, 7, 6, 0, 0, 0]

</div>

经过映射的预处理之后，原来的 `token_seq` 数据就变为了相同长度的 ID 序列，仍然对每个 ID 用一个 embedding 进行表示即可。

**Reply by kuzma-long** [reply]
感谢回复，解释的很详细，还想问一下token_seq对应的ID序列，每个ID做embedding之后，是做了平均池化吗还是concatenate？

**Reply by Sherry-XLL** [reply]
@kuzma-long 您好，如果是 context-aware 类型的模型的话，聚合 embedding 的方式可以自行选择池化的模式，比如最大池化 `max`、加和池化 `sum` 和平均池化 `mean`：

https://github.com/RUCAIBox/RecBole/blob/896155c56fb2f426554d1179dae85dc0a8f4a38e/recbole/model/abstract_recommender.py#L302-L339
